trace/tracestate/tracestate.go

```diff
@@ -0,0 +1,182 @@
+		return "", false
+	}
+	for _, entry := range ts.tracestateEntries {
+		if entry.key == key {
+			return entry.value, true
+		}
+	}
+	return "", false
+}
+
+// TraceEntries returns a slice of TracestateEntry.
+func (ts *Tracestate) TraceEntries() []*TracestateEntry {
+	return ts.tracestateEntries
+}
+
+func (ts *Tracestate) remove(key string) {
+	for index, entry := range ts.tracestateEntries {
+		if entry.key == key {
+			ts.tracestateEntries = append(ts.tracestateEntries[:index], ts.tracestateEntries[index+1:]...)
+			break
+		}
+	}
+}
+
+func (ts *Tracestate) add(entry *TracestateEntry) error {
+	ts.remove(entry.Key())
+	if len(ts.tracestateEntries) >= maxKeyValuePairs {
+		return fmt.Errorf("Set failed: reached maximum key/value pairs limit of %d", maxKeyValuePairs)
+	}
+	ts.tracestateEntries = append([]*TracestateEntry{entry}, ts.tracestateEntries...)
+	return nil
+}
+
+func isValid(key, value string) bool {
+	return keyValidationRegExp.MatchString(key) &&
+		valueValidationRegExp.MatchString(value)
+}
+
+// NewTracestateEntry creates a TracestateEntry object with given key and value.
+// It returns error if either key or value is invalid.
+func NewTracestateEntry(key, value string) (*TracestateEntry, error) {
+	if isValid(key, value) {
+		return &TracestateEntry{
+			key:   key,
+			value: value,
+		}, nil
+	}
+	return nil, errors.New("invalid parameters")
+}
+
+func containsDuplicateKey(entries []*TracestateEntry) (string, bool) {
+	keyMap := make(map[string]int)
+	for _, entry := range entries {
+		if _, ok := keyMap[entry.Key()]; ok {
+			return entry.Key(), true
+		}
+		keyMap[entry.Key()] = 1
+	}
+	return "", false
+}
+
+// NewTracestateFromEntryArray creates a Tracestate object from an array of key-value pair.
+// nil is returned with with an error if
+//  1. If the len of the entries  > maxKeyValuePairs
+//  2. If the entries contain duplicate keys
+func NewTracestateFromEntryArray(entries []*TracestateEntry) (*Tracestate, error) {
+
+	if entries == nil {
+		return nil, errors.New("Invalid parameter")
+	}
+	if len(entries) == 0 || len(entries) > maxKeyValuePairs {
+		return nil, fmt.Errorf("Invalid number of tracestateEntry (%d)", len(entries))
+	}
+	if key, duplicate := containsDuplicateKey(entries); duplicate {
+		return nil, fmt.Errorf("Contains duplicate keys (%s)", key)
+	}
+
+	tracestate := Tracestate{}
+
+	if entries != nil {
+		tracestate.tracestateEntries = append([]*TracestateEntry{}, entries...)
+	}
+
+	return &tracestate, nil
+}
+
+// NewTracestateFronParent creates a Tracestate object and adds a key-value pair to the list.
+// If a non-empty parent is passed then key/value pair from the parent is copied
+// to a newly created Tracestate object.
+// If the key already exists in the parent then its value is replaced with the
+// value passed to this function. The key is also moved to the front of
+// the list. See add func.
+func NewTracestateFronParent(parent *Tracestate, key, value string) (*Tracestate, error) {
```

Comment:
Typo in function name. Also function could be NewFromParent (remove Tracestate?)